api/vm/compile.ts

```diff
@@ -1,52 +1,84 @@
 import * as core from "@dataform/core";
+import { dataform } from "@dataform/protos";
 import * as crypto from "crypto";
 import * as fs from "fs";
 import * as os from "os";
 import * as path from "path";
 import { util } from "protobufjs";
 import { NodeVM } from "vm2";
-import { genIndex } from "../gen_index";
-
-export interface ICompileIPCParameters {
-  projectDir: string;
-  schemaSuffixOverride?: string;
-}
+import { createGenIndexConfig } from "./gen_index_config";
 
 export interface ICompileIPCResult {
   path?: string;
   err?: string;
 }
 
-export function compile(projectDir: string, schemaSuffixOverride?: string): Uint8Array {
-  const vm = new NodeVM({
+export function compile(compileConfig: dataform.ICompileConfig): Uint8Array {
+  const vmIndexFileName = path.resolve(path.join(compileConfig.projectDir, "index.js"));
+
+  const indexGeneratorVm = new NodeVM({
+    wrapper: "none",
+    require: {
+      context: "sandbox",
+      root: compileConfig.projectDir,
+      external: true
+    }
+  });
+
+  // TODO: Once all users of @dataform/core are updated to include compiler functions, remove
+  // this exception handling code (and assume existence of genIndex / compiler functions in @dataform/core).
+  const findGenIndex = (): ((base64EncodedConfig: string) => string) => {
+    try {
+      return (
+        indexGeneratorVm.run('return require("@dataform/core").genIndex', vmIndexFileName) ||
+        core.genIndex
```

Comment:
i'm confused - what do you mean by 'keep the gen index from before'?

if it helps: my intention here is to remove this dependency on core right at the point when all users have upgraded their @dataform packages past this commit - we'll rely solely on the genIndex function returned from NodeVM.

Reply:
So in that interim period, changes to core/gen_index are going to have to be backwards compatible with projects before and after this commit which might be tricky.

I guess I am saying if you want you could branch that file into 2:
- the current version that lives in api/legacy_gen_index that we use for projects before this commit
- a new version core/gen_index that we can now safely change in backwards breaking ways

It might be nice if after this commit you have confidence in changing core/* without having to think about older projects, but for now you won't be able to do that.

Reply:
Ahh, very good point, I had forgotten that of course we want to make breaking changes to the compiler inside `core` and so need to keep an old copy of the original code around to prevent breakages.